src/materials/czm/CZMMaterialBasePK1.C

```diff
@@ -0,0 +1,336 @@
+            for (unsigned int j = 0; j < 3; j++)
+              for (unsigned int k = 0; k < 3; k++)
+                for (unsigned int l = 0; l < 3; l++)
+                  R2temp(p, q) += dR_dF(i, k, p, q) * _normals[_qp](k) *
+                                      _DL_avg[_qp](i, j) * _R_avg[_qp](j, l) *
+                                      _normals[_qp](l) +
+                                  dR_dF(j, l, p, q) * _normals[_qp](l) *
+                                      _DL_avg[_qp](i, j) * _R_avg[_qp](i, k) *
+                                      _normals[_qp](k) +
+                                  dDL_dF(i, j, p, q) * _R_avg[_qp](i, k) *
+                                      _normals[_qp](k) * _R_avg[_qp](j, l) *
+                                      _normals[_qp](l);
+        }
+
+      // assemble ddsdotdS_dF*_D
+      dB_dF =
+          _D.outerProduct(dtraceDL_dF - R2temp) + _dadot_da_avg[_qp] * dC_dF;
+    } // end of area cahnges derivatives
+
+    // assemble _dPK1traction_dF as the sum of three contributions: the 1st
+    // terms  includes dadA_dF, the 2nd term includes the contributions of
+    // dB_dF, dC_dF and dD_dF, the 3rd term includes the contritbution of
+    // djump_dF
+
+    //  T1;
+    if (_use_area_change) {
+      Vtemp = (_B + _C) * _traction[_qp] + _D * _traction_inc[_qp];
+      _dPK1traction_dF[_qp] = RjkVi(da_dF, Vtemp);
+    } else
+      _dPK1traction_dF[_qp].zero();
+
+    //  T2;
+    Vtemp = _traction[_qp] + _traction_inc[_qp];
+    RankThreeTensor dT_dF_temp = _a * RijklVj(dC_dF, Vtemp);
+    if (_use_area_change)
+      dT_dF_temp += _a * RijklVj(dB_dF, _traction[_qp]);
+    _dPK1traction_dF[_qp] += dT_dF_temp;
+
+    //  T3;
+    R2temp = (_B + _C + _D) * _dtraction_djump[_qp];
+    Vtemp = _displacement_jump_global[_qp] + _displacement_jump_global_inc[_qp];
+    R3temp = RijklVi(dC_dF, Vtemp);
+    _dPK1traction_dF[_qp] += _a * RijRjkl(R2temp, R3temp);
+  }
+}
+
+void CZMMaterialBasePK1::computeFandL() {
+  RankTwoTensor F = (RankTwoTensor::Identity() +
+                     RankTwoTensor((*_grad_disp[0])[_qp], (*_grad_disp[1])[_qp],
+                                   (*_grad_disp[2])[_qp]));
+  RankTwoTensor F_neighbor = (RankTwoTensor::Identity() +
+                              RankTwoTensor((*_grad_disp_neighbor[0])[_qp],
+                                            (*_grad_disp_neighbor[1])[_qp],
+                                            (*_grad_disp_neighbor[2])[_qp]));
+
+  _F_avg[_qp] = 0.5 * (F + F_neighbor);
+  _F_avg_inv = _F_avg[_qp].inverse();
+  _DL_avg[_qp] = RankTwoTensor::Identity() - _F_avg_old[_qp] * _F_avg_inv;
+  _F_avg[_qp].getRUDecompositionRotation(_R_avg[_qp]);
+  _U_avg[_qp] = _R_avg[_qp].transpose() * _F_avg[_qp];
+  _DF_avg[_qp] = _F_avg[_qp] * _F_avg_old[_qp].inverse();
+  _DR_avg[_qp] = _R_avg[_qp] - _R_avg_old[_qp];
+  _J = _F_avg[_qp].det();
+}
+
+RankFourTensor CZMMaterialBasePK1::computedRdF(const RankTwoTensor &R,
```

Comment:
Is this a method (applies only to this class) or a function (which could be used for arbitrary tensors R and U)?

Reply:
this could definitely be a function. Ideally given a generic R2Tensor you can always find T=RU and this function is dR_dT